kb/2024-11-28/kubevirt_cert_rotation.md

```diff
@@ -0,0 +1,58 @@
+---
+title: KubeVirt Certificates Rotation
+description: KubeVirt Certificates Rotation.
+slug: kubevirt_certificates_rotation
+authors:
+  - name: Cooper Tseng
+    title: Software Engineer
+    url: https://github.com/brandboat
+    image_url: https://github.com/brandboat.png
+tags: [harvester, kubevirt, certificates, cert, ca]
+hide_table_of_contents: false
+---
+
+- Harvester: [Issue 5798](https://github.com/harvester/harvester/issues/5798)
+
+In Harvester's embedded Rancher UI, you may see some warnings on the secrets page indicating that the KubeVirt certicates are about to expire. While there is no need to worry about it as KubeVirt handles automatic certificate rotation. 
+
+![kubevirt-certs-expired](./imgs/kubevirt_certs_expired.png)
+
+This page explains how KubeVirt manages self-signed certificates, the configuration, and the triggers for certificate rotation.
+
+# KubeVirt Certificate Rotation Strategy
+KubeVirt applies self-signed certificate mechamism, where both CA and certifcates are rotated on a defined recurring interval. You can check the corresponding setting  `certificateRotateStrategy` by using cmd
+```sh
+kubectl get kubevirt -n harvester-system -o yaml
+```
+By default, the `certificateRotateStrategy` is empty, meaning KubeVirt will use its default rotation settings, and no manual configuration is needed.
+```yaml
+certificateRotateStrategy: {}
+```
+A complete configuration look like this:
+```yaml
+certificateRotateStrategy:
+  selfSigned:
+    ca:
+      duration: 24h
+      renewBefore: 4h
+    server:
+      duration: 168h
+      renewBefore: 28h
+```
+Explanation of Fields:
+- `.ca.duration`: refers to the validity period the CA certificate. The default value is 7 days.
+- `.ca.renewBefore`: refers to how much time before the CA certificate expiration a new certificate should be issued. The default value is 7 days * 0.2.
+- `.server.duration`: refers to the validity period of server certificates (e.g. virt-api, virt-handler, virt-operator, etc.). The default value is 1 day.
+- `.server.renewBefore`: refers to how much time before the server certificates expire a new certificate should be issued. The default value is 1 day * 0.2.
```

Comment:
How are the default values actually specified? I'm asking because the config example you provided uses hours while the default values all use "day" as the unit of measurement.

Reply:
My bad, the default values are
- `.ca.duration` : 168h
- `.ca.renewBefore`: 33.6h
- `.server.duration`: 24h
- `.server.renewBefore`: 4.8h

And thanks for the suggestion, the unit should all be `h` here. Behind the scene KubeVirt use [Duration](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/duration) as unit.

I'll update everything in the latest commit, thanks.